File: packages/hpack-html-dart/lib/src/encoder.dart
import 'dart:convert';
import 'dart:typed_data';
import 'format.dart';
import 'varint.dart';
import 'types.dart';

/// Encode a known header field into bytes.
Uint8List _encodeKnownField(HeaderField field) {
  final valueBytes = utf8.encode(field.value);
  final lenBytes = encodeVarInt(valueBytes.length);
  final result = BytesBuilder();
  result.addByte(field.type);
  result.add(lenBytes);
  result.add(valueBytes);
  return result.toBytes();
}

/// Encode a custom header field into bytes.
Uint8List _encodeCustomField(HeaderField field) {
  final nameBytes = utf8.encode(field.name ?? '');
  final valueBytes = utf8.encode(field.value);
  final valueLenBytes = encodeVarInt(valueBytes.length);
  final result = BytesBuilder();
  result.addByte(field.type);
  result.addByte(nameBytes.length);
  result.add(nameBytes);
  result.add(valueLenBytes);
  result.add(valueBytes);
  return result.toBytes();
}

/// Encode a header field.
Uint8List _encodeField(HeaderField field) {
  return field.type >= fieldCustomStart
      ? _encodeCustomField(field)
      : _encodeKnownField(field);
}

/// Options for encoding a packet.
class EncodePacketOptions {
  final int flags;
  final List<HeaderField> fields;
  final Uint8List compressedBody;
  final int uncompressedLength;
  final int? crc32Value;

  const EncodePacketOptions({
    required this.flags,
    required this.fields,
    required this.compressedBody,
    required this.uncompressedLength,
    this.crc32Value,
  });
}

/// Encode a complete .hpack packet.
Uint8List encodePacket(EncodePacketOptions options) {
  final encodedFields = options.fields.map(_encodeField).toList();
  final fieldCountBytes = encodeVarInt(options.fields.length);

  final headerContentLength =
      fieldCountBytes.length + encodedFields.fold<int>(0, (sum, f) => sum + f.length);
  final headerLenBytes = encodeVarInt(headerContentLength);

  final uncompressedLenBytes = encodeVarInt(options.uncompressedLength);
  final hasCrc = options.crc32Value != null;

  final packet = BytesBuilder();

  // Magic
  packet.add(magic);
  // Version
  packet.addByte(version);
  // Flags
  packet.addByte(options.flags);
  // Header length
  packet.add(headerLenBytes);
  // Field count
  packet.add(fieldCountBytes);
  // Fields
  for (final f in encodedFields) {
    packet.add(f);
  }
  // Uncompressed length
  packet.add(uncompressedLenBytes);
  // Compressed body
  packet.add(options.compressedBody);
  // CRC32 (little-endian)
  if (hasCrc) {
    final crc = options.crc32Value!;
    packet.addByte(crc & 0xFF);
    packet.addByte((crc >> 8) & 0xFF);
    packet.addByte((crc >> 16) & 0xFF);
    packet.addByte((crc >> 24) & 0xFF);
  }

  return packet.toBytes();
}

File: packages/hpack-html/src/encoder.ts
/**
 * .hpack packet encoder.
 *
 * Serializes metadata headers + compressed body into the binary .hpack format.
 */

import { encodeVarInt } from "./varint.js";
import {
  MAGIC,
  VERSION,
  FIELD_CUSTOM_START,
  CRC32_SIZE,
} from "./format.js";
import type { HeaderField } from "./types.js";

/**
 * Encode a single known header field (type < 0x10) into bytes.
 *
 * Layout: [fieldType: 1B] [valueLength: VarInt] [value: UTF-8]
 */
function encodeKnownField(field: HeaderField): Uint8Array {
  const valueBytes = new TextEncoder().encode(field.value);
  const lenBytes = encodeVarInt(valueBytes.length);

  const result = new Uint8Array(1 + lenBytes.length + valueBytes.length);
  result[0] = field.type;
  result.set(lenBytes, 1);
  result.set(valueBytes, 1 + lenBytes.length);
  return result;
}

/**
 * Encode a custom header field (type >= 0x10) into bytes.
 *
 * Layout: [fieldType: 1B] [nameLength: 1B] [name: UTF-8] [valueLength: VarInt] [value: UTF-8]
 */
function encodeCustomField(field: HeaderField): Uint8Array {
  const nameBytes = new TextEncoder().encode(field.name ?? "");
  const valueBytes = new TextEncoder().encode(field.value);
  const valueLenBytes = encodeVarInt(valueBytes.length);

  const result = new Uint8Array(
    1 + 1 + nameBytes.length + valueLenBytes.length + valueBytes.length,
  );
  let offset = 0;
  result[offset++] = field.type;
  result[offset++] = nameBytes.length;
  result.set(nameBytes, offset);
  offset += nameBytes.length;
  result.set(valueLenBytes, offset);
  offset += valueLenBytes.length;
  result.set(valueBytes, offset);
  return result;
}

/**
 * Encode a header field (dispatches to known or custom encoding).
 */
function encodeField(field: HeaderField): Uint8Array {
  if (field.type >= FIELD_CUSTOM_START) {
    return encodeCustomField(field);
  }
  return encodeKnownField(field);
}

export interface EncodePacketOptions {
  /** Flags byte (already encoded via encodeFlags). */
  flags: number;
  /** Header fields to include. */
  fields: HeaderField[];
  /** Compressed body bytes. */
  compressedBody: Uint8Array;
  /** Original uncompressed body length in bytes. */
  uncompressedLength: number;
  /** CRC32 checksum of uncompressed body (included only if flags has checksum bit). */
  crc32?: number;
}

/**
 * Encode a complete .hpack packet.
 *
 * Layout:
 *   [magic: 4B] [version: 1B] [flags: 1B]
 *   [headerLength: VarInt] [fieldCount: VarInt] [fields...]
 *   [uncompressedLength: VarInt] [compressedBody...] [crc32?: 4B LE]
 */
export function encodePacket(options: EncodePacketOptions): Uint8Array {
  const { flags, fields, compressedBody, uncompressedLength, crc32 } = options;

  // Encode all fields
  const encodedFields: Uint8Array[] = fields.map(encodeField);
  const fieldCountBytes = encodeVarInt(fields.length);

  // Header section content = fieldCount + all encoded fields
  const headerContentLength =
    fieldCountBytes.length +
    encodedFields.reduce((sum, f) => sum + f.length, 0);
  const headerLenBytes = encodeVarInt(headerContentLength);

  // Body section
  const uncompressedLenBytes = encodeVarInt(uncompressedLength);
  const hasCrc = crc32 !== undefined;
  const crcSize = hasCrc ? CRC32_SIZE : 0;

  // Total packet size
  const totalSize =
    MAGIC.length +          // magic
    1 +                     // version
    1 +                     // flags
    headerLenBytes.length + // header length varint
    headerContentLength +   // header content
    uncompressedLenBytes.length + // uncompressed length varint
    compressedBody.length + // compressed body
    crcSize;                // optional CRC32

  const packet = new Uint8Array(totalSize);
  let offset = 0;

  // Magic
  packet.set(MAGIC, offset);
  offset += MAGIC.length;

  // Version
  packet[offset++] = VERSION;

  // Flags
  packet[offset++] = flags;

  // Header length
  packet.set(headerLenBytes, offset);
  offset += headerLenBytes.length;

  // Field count
  packet.set(fieldCountBytes, offset);
  offset += fieldCountBytes.length;

  // Fields
  for (const encoded of encodedFields) {
    packet.set(encoded, offset);
    offset += encoded.length;
  }

  // Uncompressed length
  packet.set(uncompressedLenBytes, offset);
  offset += uncompressedLenBytes.length;

  // Compressed body
  packet.set(compressedBody, offset);
  offset += compressedBody.length;

  // CRC32 (little-endian)
  if (hasCrc) {
    packet[offset++] = crc32 & 0xff;
    packet[offset++] = (crc32 >>> 8) & 0xff;
    packet[offset++] = (crc32 >>> 16) & 0xff;
    packet[offset++] = (crc32 >>> 24) & 0xff;
  }

  return packet;
}
